{% extends 'umi/base.html' %}

{% block content %}

{% load static %}
<title>Is There an Amplification Bias?</title>
<header class="w3-container w3-theme w3-padding" id="myHeader">
  <div class="w3-center">
  <h1 class="w3-xxxlarge w3-animate-bottom">Are all UMIs amplified at the same rate?</h1>
</div>
<script>
document.body.classList.add('js-loading');

window.addEventListener("load", showPage);

function showPage() {
  document.body.classList.remove('js-loading');
}
</script>
</header>
<br>

<div class="w3-container w3-justify">
  <p>Another aspect to UMIs that bears scrutiny is their relationship to amplification bias during PCR. Although the UMI tag is small compared to the rest of the transcript it is attached to, it still may somehow play a role in the amplification reaction. Finding that UMIs do exert some kind of amplification bias would result in mixed implications for the technique. Their purpose is to negate these biases through molecular counting, so notionally, introducing biases of their own is counterproductive but ultimately not significant. However, when combined with the previous evidence that UMIs are not as random—and thus not as effective at nullifying PCR biases—as previously thought, there is a potential for these effects to substantially impact the results.</p>
  <p>This page introduces the concept of sequencing reads as another measurement of UMIs. Previously, UMI counts have been discussed. These counts, in theory, indicate the number of molecular instances of a UMI prior to any amplification. The number of reads of a UMI is the number of times it was sequenced and should be representative of the number of UMI instances after amplification. The more molecules, the more they will be sequenced (depending on the sequencing depth of the experiment). Reads can be thought of as a representative, but not exact, figure for a UMI’s count after amplification. Another new value, here called the PCR amplification factor (PAF), is the total number of reads of a UMI divided by its count. In essence, the PAF for a UMI is the extent to which it has been amplified. In other words, it is the mean number of reads across all instances of a UMI.</p>
</div>
<br>

<div class='w3-container'>
  <button class="w3-button w3-block w3-theme w3-hover-bar-theme w3-left-align accordion">Comparison of UMI reads across two experiments</button>
  <div id="reads2exp" class="w3-container panel w3-justify">
    <br>
    <!-- <canvas id="reads2expChart"></canvas> -->
    <img src="{%  static 'umi/images/ampBias1.png' %}" style="width:100%" alt="UMI Count 2 Experiments">
    <p>As a first step, UMI reads were compared across two experiments to see if they are consistent. It was already known that UMI counts correlate across these two experiments, so this result is largely as expected. The reason experiment 2 has higher reads in many instances is due to the fact that it has more reads overall than experiment 1.</p>
  </div>
  <button class="w3-button w3-block w3-theme w3-hover-bar-theme w3-left-align accordion">Comparison of a UMI's reads and its frequency</button>
  <div id="readsVCounts" class="w3-container panel w3-justify">
    <br>
    <!-- <canvas id="readsVCountsChart"></canvas> -->
    <img src="{%  static 'umi/images/ampBias2.png' %}" style="width:100%" alt="UMI Count 2 Experiments">
    <p>This chart compares a UMI's reads with its count, which should be highly correlated. If all UMIs are amplified equally, then the UMI that appears most frequently before amplification (count) should also be most frequent after (reads). The result is that while these two values do correlate to some extent, there is still a fair amount of variation in the amplification that is not explained by a UMI's count. This is most evidently seen at the low end of the count scale where many UMIs have the same count value but a large variation in the number of reads. </p>
  </div>
  <button class="w3-button w3-block w3-theme w3-hover-bar-theme w3-left-align accordion">PAF compared with UMI frequency</button>
  <div id="readsPerVCount" class="w3-container panel w3-justify">
    <br>
    <!-- <canvas id="readsPerVCountChart"></canvas> -->
    <img src="{%  static 'umi/images/ampBias3.png' %}" style="width:100%" alt="UMI Count 2 Experiments">
    <p>Here, the PCR amplification factor for each UMI is compared with its count to see if UMIs of a certain frequency are being amplified more than others. The lack of a pattern suggests that this is not the case and that PAF varies across the spectrum of UMI frequency. Like the previous chart, there appears to be a large amount of variation in amplification. This in itself does not indicate any bias, as it could simply be the result of random effects during amplification.</p>
    <!-- <div class="slidecontainer">
      <input type="range" min="10" max="1000" value="500" class="slider" step="1" id="myRange">
    </div>
    <div id='demo'></div> -->

  </div>
  <button class="w3-button w3-block w3-theme w3-hover-bar-theme w3-left-align accordion">PAF across two experiments</button>
  <div id="readsPer2exp" class="w3-container panel w3-justify">
    <br>
    <!-- <canvas id="readsPer2expChart"></canvas> -->
    <img src="{%  static 'umi/images/ampBias4.png' %}" style="width:100%" alt="UMI Count 2 Experiments">
    <p>This chart provides a very interesting and surprising insight. The same UMIs are being amplified more in both experiments with a strong correlation. This result suggests that something about these UMIs makes them consistently biased in favour of amplification.</p>
  </div>
  <button class="w3-button w3-block w3-theme w3-hover-bar-theme w3-left-align accordion">Details of the highest PAF UMIs</button>
  <div id="pafDetail" class="w3-container panel w3-justify">
    <br>
    <canvas id="pafDetailChart"></canvas>
    <p>Unlike the UMIs with the highest pre-amplification count, the top PAF UMIs do not share any obvious sequence similarities. There is no pattern of highly repetitive sequences and in fact, none of the top 50 UMIs by count appear in this chart. There seems to be some factor other than sequence that determines the PAF of a UMI.</p>
  </div>
  <button class="w3-button w3-block w3-theme w3-hover-bar-theme w3-left-align accordion">Lorenz curve and amplification inequality</button>
  <div id="gini" class="w3-container panel w3-justify">
    <canvas id="giniChart"></canvas>
    <p>The Lorenz curve is commonly used in economics as a representation of wealth or income inequality in a society. Here, it is being used to measure the inequality of amplification 'wealth' among the population of UMIs in the data set. The chart can be interpreted as follows: the bottom 80% of all UMIs have about 22% of the amplification 'wealth,' which is measured in PAF. In other words, after accounting for frequency, 78% of all reads are of a select 20% of UMIs. This inequality is reflected in the Gini coefficient, 73.8, which is worse than income inequality in South Africa, according to The World Factbook<sup>4</sup>.</p>
  </div>
</div>

<br>

<div class="w3-container w3-justify">
  <h5>Discussion</h5>
  <p>The results of these analyses clearly show that there is some kind of amplification bias present in this data set. There is an unequal distribution of amplification across UMIs and certain UMIs are consistently amplified more in multiple experiments. The most puzzling finding is that there is no pattern among the sequences of these highly amplified UMIs, which precludes an easy explanation for their behaviour. However, perhaps it is logical that the sequence itself is not the cause given the small size of UMIs compared to the attached transcript. If there is a molecular basis for this amplification bias then it is reasonable that it would be found in the far larger and more influential sequence of the transcript. This topic is the subject of the next page.</p>
  <p></p>
</div>
<div class="w3-row w3-container">
    <a class="w3-btn w3-theme w3-hover-theme w3-left buttL" href="/umi/umi-random"><span>Previous</span><br>Are UMIs Random?</a>
    <a class="w3-btn w3-theme w3-hover-theme w3-right butt" href="/umi/gene-bias"><span>Next</span><br>Is There a Gene Bias?</a>
</div>

<br>

<script>
var acc = document.getElementsByClassName("accordion");
var w;

for (w = 0; w < acc.length; w++) {
  acc[w].addEventListener("click", function() {
    this.classList.toggle("active");
    var panel = this.nextElementSibling;
    if (panel.style.maxHeight){
      panel.style.maxHeight = null;
      panel.style.border = null;
    } else {
      panel.style.maxHeight = panel.scrollHeight + "px";
      panel.style.border = 1 + "px solid #b7d9fc";
    }
  });
}

// var ctx2 = document.getElementById('reads2expChart').getContext('2d');
// // var smallx = ampBias_test_x.slice(0, 500)
// // var smally = ampBias_test_y.slice(0, 500)
// var readsTwoExpData = ampBias_1_xcounts.map((x, i) => {
//   return {
//     x: x,
//     y: ampBias_1_ycounts[i]
//   };
// });

// var chart2 = new Chart(ctx2, {
//     // The type of chart we want to create
//     type: 'scatter',

//     // The data for our dataset
//     data: {
//         labels: umiRandom_1_umis,
//         datasets: [{
//             label: "Stuff",
//             data: readsTwoExpData,
//             showLine: false,
//             pointBackgroundColor: 'rgb(99, 99, 255, 0.1)',
//             pointBorderColor: 'rgb(99, 99, 255, 0.2)',
//             pointRadius: 4,
//         }],
//     },

//     // Configuration options go here
//     options: {
//         animation: false,
//         legend: {
//         display: false,
//       },
//         tooltips: {
//           enabled: false
//         },
//         scales: {
//             yAxes: [{
//                 scaleLabel: {
//                     display: true,
//                     fontSize: 17,
//                     labelString: 'Log UMI Reads in Experiment 2'
//                 }
//             }],
//             xAxes: [{
//                 scaleLabel: {
//                     display: true,
//                     fontSize: 17,
//                     labelString: 'Log UMI Reads in Experiment 1'
//                 }
//             }]
//         }
//     }
// });

// var ctx3 = document.getElementById('readsVCountsChart').getContext('2d');
// // var smallx = ampBias_test_x.slice(0, 500)
// // var smally = ampBias_test_y.slice(0, 500)
// var readsVCountsData = ampBias_2_counts.map((x, i) => {
//   return {
//     x: x,
//     y: ampBias_2_reads[i]
//   };
// });

// var chart3 = new Chart(ctx3, {
//     // The type of chart we want to create
//     type: 'scatter',

//     // The data for our dataset
//     data: {
//         labels: umiRandom_1_umis,
//         datasets: [{
//             label: "Stuff",
//             data: readsVCountsData,
//             showLine: false,
//             pointBackgroundColor: 'rgb(99, 99, 255, 0.1)',
//             pointBorderColor: 'rgb(99, 99, 255, 0.2)',
//             pointRadius: 4,
//         }],
//     },

//     // Configuration options go here
//     options: {
//         animation: false,
//         legend: {
//         display: false,
//       },
//         tooltips: {
//           enabled: false
//         },
//         scales: {
//             yAxes: [{
//                 scaleLabel: {
//                     display: true,
//                     fontSize: 17,
//                     labelString: 'Log UMI Reads'
//                 }
//             }],
//             xAxes: [{
//                 scaleLabel: {
//                     display: true,
//                     fontSize: 17,
//                     labelString: 'Log UMI frequency (count)'
//                 }
//             }]
//         }
//     }
// });
// var ctx = document.getElementById('readsPerVCountChart').getContext('2d');
// var data = ampBias_3_readsPer.map((x, i) => {
//   return {
//     x: x,
//     y: ampBias_3_count[i]
//   };
// });

// var chart = new Chart(ctx, {
//     // The type of chart we want to create
//     type: 'scatter',

//     // The data for our dataset
//     data: {
//         labels: umiRandom_1_umis,
//         datasets: [{
//             label: "Stuff",
//             data: data,
//             showLine: false,
//             pointBackgroundColor: 'rgb(99, 99, 255, 0.1)',
//             pointBorderColor: 'rgb(99, 99, 255, 0.2)',
//             pointRadius: 4,
//         }],
//     },

//     // Configuration options go here
//     options: {
//         animation: false,
//         legend: {
//         display: false,
//       },
//         tooltips: {
//           enabled: false
//         },
//         scales: {
//             yAxes: [{
//                 scaleLabel: {
//                     display: true,
//                     fontSize: 17,
//                     labelString: 'Log PCR Amplification Factor'
//                 }
//             }],
//             xAxes: [{
//                 scaleLabel: {
//                     display: true,
//                     fontSize: 17,
//                     labelString: 'Log UMI frequency (count)'
//                 }
//             }]
//         }
//     }
// });

// var ctx4 = document.getElementById('readsPer2expChart').getContext('2d');
// var data4 = ampBias_4_readsPerx.map((x, i) => {
//   return {
//     x: x,
//     y: ampBias_4_readsPery[i]
//   };
// });

// var chart4 = new Chart(ctx4, {
//     // The type of chart we want to create
//     type: 'scatter',

//     // The data for our dataset
//     data: {
//         labels: umiRandom_1_umis,
//         datasets: [{
//             label: "Stuff",
//             data: data4,
//             showLine: false,
//             pointBackgroundColor: 'rgb(99, 99, 255, 0.1)',
//             pointBorderColor: 'rgb(99, 99, 255, 0.2)',
//             pointRadius: 4,
//         }],
//     },

//     // Configuration options go here
//     options: {
//         animation: false,
//         legend: {
//         display: false,
//       },
//         tooltips: {
//           enabled: false
//         },
//         scales: {
//             yAxes: [{
//                 scaleLabel: {
//                     display: true,
//                     fontSize: 17,
//                     labelString: 'Log PAF in Experiment 2'
//                 }
//             }],
//             xAxes: [{
//                 scaleLabel: {
//                     display: true,
//                     fontSize: 17,
//                     labelString: 'Log PAF in Experiment 1'
//                 }
//             }]
//         }
//     }
// });

var ctx5 = document.getElementById('giniChart').getContext('2d');
var data51 = ampBias_5_bins.map((x, i) => {
  return {
    x: x,
    y: ampBias_5_result[i]
  };
});
var data52 = ampBias_5_bins.map((x, i) => {
  return {
    x: x,
    y: ampBias_5_bins[i]
  };
});
var chart5 = new Chart(ctx5, {
    // The type of chart we want to create
    type: 'scatter',

    // The data for our dataset
    data: {
        datasets: [{
            label: "Observed",
            data: data51,
            fill: false,
            borderColor: 'rgb(255, 116, 0)'
          },
          {
            label: "Perfect Equality",
            data: data52,
            fill: false,
            borderColor: 'rgb(28, 85, 255)'
          },
        ],
    },

    // Configuration options go here
    options: {
        animation: false,
        scales: {
            yAxes: [{
                scaleLabel: {
                    display: true,
                    fontSize: 15,
                    labelString: 'Percentage of Sequencing "Wealth"'
                }
            }],
            xAxes: [{
                scaleLabel: {
                    display: true,
                    fontSize: 15,
                    labelString: 'Percentage of UMI Population'
                }
            }]
        }
    }
});

var ctx6 = document.getElementById('pafDetailChart').getContext('2d');
var chart6 = new Chart(ctx6, {
    // The type of chart we want to create
    type: 'bar',

    // The data for our dataset
    data: {
        labels: ampBias_6_umi,
        datasets: [{
            label: "PAF",
            backgroundColor: 'rgb(99, 99, 255)',
            borderColor: 'rgb(99, 99, 255)',
            data: ampBias_6_paf,
        }],

    },

    // Configuration options go here
    options: {
        scales: {
            yAxes: [{
                scaleLabel: {
                    display: true,
                    fontSize: 15,
                    labelString: 'PCR Amplification Factor'
                }
            }],
            xAxes: [{
                scaleLabel: {
                    display: true,
                    fontSize: 15,
                    labelString: 'UMI Sequence'
                },
                ticks: {
                  fontSize: 12,
                  maxRotation: 90,
                  minRotation: 90
                }
            }]
        }
    }
});

// var slider = document.getElementById("myRange");
// var output = document.getElementById("demo");
// output.innerHTML = slider.value; // Display the default slider value
//
// // Update the current slider value (each time you drag the slider handle)
// slider.oninput = function() {
//     output.innerHTML = this.value;
//     smallx = ampBias_test_x.slice(0, this.value)
//     smally = ampBias_test_y.slice(0, this.value)
//     data = smallx.map((x, i) => {
//       return {
//         x: x,
//         y: smally[i]
//       };
//     });
//     chart.data.datasets.forEach((dataset) => {
//       dataset.data = data
//     });
//     chart.update();
// }
</script>
{% endblock %}
