{% extends 'umi/base.html' %}
{% load staticfiles%}
{% block content %}
<title>UMIs in scRNA-Seq Data</title>
  <header class="w3-container w3-theme w3-padding" id="myHeader">
    <div class="w3-center">
    <h4>Are they good? How good are they? Let's find out!</h4>
    <h1 class="w3-xxxlarge w3-animate-bottom">UMIs in scRNA-Seq Data</h1>
  </div>
  </header>
  <br>
  <div>
  <a class="w3-btn w3-xlarge w3-theme-d1 w3-hover-theme" href="/umi/test">More fancy stuff</a>
  </div>
  <br>

  <div>
    <div class="w3-container">
      <h5>Introduction</h5>
      <p>Single-cell RNA sequencing (scRNA-seq) is an important tool in bioinformatics and has been applied in studies ranging from tumours to early embryonic development. Rather than the standard, relative measures of expression such as reads per kilobase million (RPKM), molecular counting through the use of unique molecular identifiers (UMIs) has recently gained traction in the field. UMIs provide an absolute scale of measurement that can be intuitively compared across experiments as well as the means to capture more elusive transcripts that previously escaped detection. Because all current scRNA-seq methods involve amplification, the ability of UMIs to neutralise biases introduced by PCR is vital<sup>1,2</sup>.</p>
      <img src="{%  static 'umi/images/UMIweb.png' %}" style="width:100%" alt="UMI Diagram">
      <p>UMIs are 5- to 10-base-long sequences that are appended to mRNA transcripts that have been extracted from a cell. Although short, they can achieve a high level of variability because at each position, one of four nucleotide residues can exist. The figure above shows a simplified example of a UMI scRNA-seq procedure. In <b>a</b>, transcripts of three genes (indicated by the colours) are extracted from an orange cell. Black UMIs are appended to these transcripts in <b>b</b>, with the variable UMI sequences indicated by different shapes. The transcripts are amplified to approximately the same degree in <b>c</b>. Finally, in <b>d</b> the result is sequenced and analysed computationally. By counting unique combinations of gene and UMI, it is possible to derive the exact number of transcripts of each gene that was present in the cell<sup>1</sup>.</p>
      <p>The utility of UMIs depends on their randomness and variability, as this is what prevents situations where two of the same UMI bind transcripts of the same gene<sup>1</sup>. In that scenario, one of the transcripts would effectively be lost because it would appear to have come from amplification and not the cell itself. Alarmingly, observations from a recent experiment with <i>Schizosaccharomyces pombe</i> (fission yeast) cells seem to indicate that UMIs are not as random as previously assumed. This report was designed to investigate the behaviour of UMIs in this data set as well as the role of amplification bias and whether certain genes were bound preferentially. Information about data sources and experimental details can be found on the <a href='about/'>About</a> page.</p>

      <!-- <canvas id="myChart"></canvas> -->
    </div>
    <div class="w3-container">
      <h5>Navigating This Site</h5>
      <p>This website is divided into pages centered around different aspects of the study. Click the <b>Start</b> button below to advance to the next page. Charts and other supporting data are hidden in tabs like the one below to avoid an overload of information. Simply click on the tab to make its contents appear or disappear.</p>
    </div>
  </div>
  <br>
  <div class='w3-container'>
    <button class="w3-button w3-block w3-theme w3-hover-bar-theme w3-left-align accordion">Click me!</button>
    <div id="example" class="w3-container panel">
        <canvas id="exampleChart"></canvas>
        <p>Below, you will find a short description of the chart and a brief discussion of its implications.</p>

    </div>
  </div>
  <br>
  <div class="w3-container">
    <a class="w3-btn w3-theme w3-hover-theme w3-right" href="/umi/umi-random">Start <i class='fa fa-arrow-right'></i><br> Are UMIs Random?</a>
  </div>

  <br>
  <br>

<script>
var acc = document.getElementsByClassName("accordion");
var w;

for (w = 0; w < acc.length; w++) {
  acc[w].addEventListener("click", function() {
    this.classList.toggle("active");
    var panel = this.nextElementSibling;
    if (panel.style.maxHeight){
      panel.style.maxHeight = null;
      panel.style.border = null;
    } else {
      panel.style.maxHeight = panel.scrollHeight + "px";
      panel.style.border = 1 + "px solid #b7d9fc";
    }
  });
}
var ctx = document.getElementById('exampleChart').getContext('2d');
var chart = new Chart(ctx, {
    // The type of chart we want to create
    type: 'line',

    // The data for our dataset
    data: {
        labels: ['','','','','','','','','','','','','','','',''],
        datasets: [{
            label: "Bioinformatics Knowledge",
            borderColor: 'rgb(255, 116, 0)',
            fill: false,
            data: [20, 25, 30, 35, 40, 45, 50, 55, 60, 63, 65, 67, 69, 70, 71, 72],
        },
        {
            label: "Web Development Knowledge",
            borderColor: 'rgb(28, 85, 255)',
            fill: false,
            data: [0, 0, 0, 0, 0, 0, 0, 5, 20, 40, 60, 80, 95, 100, 100, 100],
        }],

    },

    // Configuration options go here
    options: {
        scales: {
            yAxes: [{
                scaleLabel: {
                    display: false,
                    labelString: 'Count'
                }
            }],
            xAxes: [{
                scaleLabel: {
                    display: false,
                    labelString: 'UMI'
                },
                ticks: {
                  fontSize: 12,
                  maxRotation: 90,
                  minRotation: 90
                }
            }]
        }
    }
});
</script>
{% endblock %}
