{% extends 'umi/base.html' %}

{% block content %}

<title>Mammalian Data</title>
<header class="w3-container w3-theme w3-padding" id="myHeader">
  <div class="w3-center">
  <h1 class="w3-xxxlarge w3-animate-bottom">How do UMIs behave in mammalian samples?</h1>
</div>
</header>
<br>

<div class="w3-container w3-justify">
  <p>The discoveries up to this point have all come from data originating in <i>S. pombe</i> cells. Many of the applications of scRNA-seq such as the study of early development and cellular disease states involve mouse or human cells. While the fundamental elements of the technique are the same, it is prudent to try to replicate the previous findings using data from mammalian cells. The data in this section comes from a sample of 100 human cells and is publicly available online. As always, there is a more detailed description of the data on the <a href='/umi/about'>About</a> page.</p>
  <p>It is important to note that this data uses UMIs that are 10 bases long instead of 8 and that it originated from a different lab that may use different protocols. Due to the larger length of the sequence, the UMIs in this data set have many more (4<sup>10</sup> = 1,048,576) possible permutations and so the probability of two UMIs occuring with the same sequence is much lower. In addition, the sequencing depth of this experiment is lower, which results in a lower number of reads per cell and lower read counts overall across both UMIs and genes. Naturally, this is not ideal for comparison with the <i>S. pombe</i> data, but the basic first steps of analysis can be carried out.</p>
  <p></p>
</div>

<div class='w3-container'>
  <button class="w3-button w3-block w3-theme w3-hover-bar-theme w3-left-align accordion">Histogram of UMI frequencies</button>
  <div id="histo" class="w3-container panel w3-justify">
      <canvas id="histoChart"></canvas>
      <br>
      <div class="slidecontainer">
        <input type="range" min="10" max="280000" value="130" class="slider" step="1" id="myRange">
      </div>
      <div id='demo'></div>
      <p>Like the previous histogram, the y axis can be adjusted using the slider to reveal the full scale of this chart. Notice that in this data set the most frequent UMI occurs only 85 times, smaller than the maximum in the <i>S. pombe</i> data by a factor of about 100. Again, this is due to both a decreased read count in this experiment and the increased UMI length. Overall, this histogram shows that the human UMIs have a smaller range in their frequencies and a much shorter&mdash;but still present&mdash;tail. </p>
  </div>
  <button class="w3-button w3-block w3-theme w3-hover-bar-theme w3-left-align accordion">Frequencies of the 50 most common UMIs</button>
  <div id="Freqs" class="w3-container panel w3-justify">
      <canvas id="freqChart"></canvas>
      <p>These are the 50 most frequent UMIs in the human data. Paralleling the <i>S. pombe</i> data, the most common UMI is a simple run of the same nucleotide. However in this case, the nucleotide is thymine and not guanosine. 5 of the top 6 UMIs contain a large run of thymines, although the rest of the top 50 show much more variety than those of the yeast data. As before, the next tab goes further into detail examining the top 1000 UMIs.</p>
  </div>
  <button class="w3-button w3-block w3-theme w3-hover-bar-theme w3-left-align accordion">Composition of the 1,000 most common UMIs</button>
  <div id="Top1000" class="w3-container panel w3-justify">
    <br>
    <table class="w3-table w3-bordered">
    <thead>
    <tr class="w3-theme-l1 table-hover">
      <th></th>
      <th colspan="2" class='w3-center'>>=6 Bases Identical</th>
      <th colspan="2" class='w3-center'>>=7 Bases Identical</th>
      <th colspan="2" class='w3-center'>>=8 Bases Identical</th>
      <th colspan="2" class='w3-center'>>=9 Bases Identical</th>
    </tr>
    </thead>
    <tbody>
    <tr class='w3-theme-l1'>
      <th>Repeated Base</th>
      <th>No. UMIs</th>
      <th>Percentage</th>
      <th>No. UMIs</th>
      <th>Percentage</th>
      <th>No. UMIs</th>
      <th>Percentage</th>
      <th>No. UMIs</th>
      <th>Percentage</th>
    </tr>
    <tr>
      <td>A</td>
      <td>52</td>
      <td>35%</td>
      <td>9</td>
      <td>19%</td>
      <td>0</td>
      <td>0%</td>
      <td>0</td>
      <td>0%</td>
    </tr>
    <tr class="w3-theme-l4">
      <td>C</td>
      <td>12</td>
      <td>8%</td>
      <td>2</td>
      <td>4%</td>
      <td>0</td>
      <td>0%</td>
      <td>0</td>
      <td>0%</td>
    </tr>
    <tr>
      <td>G</td>
      <td>41</td>
      <td>28%</td>
      <td>9</td>
      <td>19%</td>
      <td>2</td>
      <td>11%</td>
      <td>0</td>
      <td>0%</td>
    </tr>
    <tr class="w3-theme-l4">
      <td>T</td>
      <td>43</td>
      <td>29%</td>
      <td>28</td>
      <td>58%</td>
      <td>16</td>
      <td>89%</td>
      <td>7</td>
      <td>100%</td>
    </tr>
    <tr class="w3-theme-l1">
      <td><b>Total</b></td>
      <td><b>148</b></td>
      <td><b>100%</b></td>
      <td><b>48</b></td>
      <td><b>100%</b></td>
      <td><b>18</b></td>
      <td><b>100%</b></td>
      <td><b>7</b></td>
      <td><b>100%</b></td>
    </tr>
    </tbody>
    </table>
    <br>
    <p>Examining the composition of the top 1000 human UMIs, it appears that there is much more variety in the sequences. Only about 15% of the top 1000 have sequences where more than half of nucleotides are identical (cf. 70% in the <i>S. pombe</i> data). As seen in the previous tab, it appears that thymine, rather than guanosine, is the most dominant nucleotide in this data. Curiously, cytosine is once again the least common nucleotide.</p>
  </div>
  <button class="w3-button w3-block w3-theme w3-hover-bar-theme w3-left-align accordion">Lorenz curve for genes</button>
  <div id="gini" class="w3-container panel w3-justify">
    <canvas id="giniChart"></canvas>
    <p>Switching focus to mammalian genes, the Lorenze curve for their PAF equality indicates that they are actually very equal. The Gini coefficient of <b>SOMETHING</b> shows that they are almost perfectly equal. This is a somewhat strange result and the likely cause as well as the reason behind foregoing a UMI Lorenz curve will be explained in the Discussion below.</p>
  </div>
</div>

<br>
<div class="w3-container w3-justify">
  <h5>Discussion</h5>
  <p>At a first glance, it appears that the mammalian data has far less bias in it than the fission yeast data set. However, the lower overall read count in this data is a strong confounding factor in all of these analyses. It is apparent that the charts featuring PAF values are conspicuously absent from this page, and there is a reason for that. Due to the low read counts, all of the PAF values for both UMIs and genes in this data set are between 1 and 2. The values of 2 come from UMIs or genes that were counted once and read twice by chance. For every other instance, they were never read more than twice the number of times they were counted. This results in such a small range of values that the Lorenz curve is barely applicable (and not at all in the case of UMIs).</p>
  <p>So unfortunately, the main result of this page is that this mammalian data is unfit for these analyses. It was used mainly for timeliness and to give some indication of how this report's findings apply to experiments using mammalian cells. In the latter regard, it was still somewhat successful in that it showed that repetitive UMIs are the most common in mammals as well. A UMI having a sequence of entirely thymines is quite literally a one-in-a-million chance and for this to be the most common UMI, there must be a bias at work. Recall that the poly(A) tail is positioned close to where UMIs anneal to the transcript (depending on the protocol)<sup>3</sup>. It is possible that thymine-rich UMIs preferentially bind the long run of adenosines and as a result, are the most common UMIs in the data.</p>
</div>
<div class="w3-row w3-container">
    <a class="w3-btn w3-theme w3-hover-theme w3-left buttL" href="/umi/gene-bias"><span>Previous</span><br>Is There a Gene Bias?</a>
    <a class="w3-btn w3-theme w3-hover-theme w3-right butt" href="/umi/conclusion"><span>Next</span><br>Conclusion</a>
</div>

<br>

<script>
var acc = document.getElementsByClassName("accordion");
var w;

for (w = 0; w < acc.length; w++) {
  acc[w].addEventListener("click", function() {
    this.classList.toggle("active");
    var panel = this.nextElementSibling;
    if (panel.style.maxHeight){
      panel.style.maxHeight = null;
      panel.style.border = null;
    } else {
      panel.style.maxHeight = panel.scrollHeight + "px";
      panel.style.border = 1 + "px solid #b7d9fc";
    }
  });
}

function histogram(data, size) {
    var length = data.length;

    var min = data[0];
    var max = data[1];

    for (var i = 0; i < length; i++) {
        var item = data[i];
        if (item < min) min = item;
        else if (item > max) max = item;
    }

    var bins = Math.ceil((max - min + 1) / size);

    var histogram = new Array(bins);

    for (var i = 0; i < bins; i++) histogram[i] = 0;

    for (var i = 0; i < length; i++)
        histogram[Math.floor((data[i] - min) / size)]++;

    return histogram;
}

var histo = histogram(mamData_1_counts, 1)
var maximum = 86
var histoBins = []
var i = 1
while (i <= maximum) {
  histoBins.push(i)
  i = i + 1
}

var ctx1 = document.getElementById('histoChart').getContext('2d');
var chart1 = new Chart(ctx1, {
    // The type of chart we want to create
    type: 'bar',

    // The data for our dataset
    data: {
        labels: histoBins,
        datasets: [{
            label: "Occurence",
            data: histo,
            backgroundColor: 'rgb(99, 99, 255)',
            borderColor: 'rgb(99, 99, 255)',
        }],

    },

    // Configuration options go here
    options: {
        scales: {
            yAxes: [{
                scaleLabel: {
                    display: true,
                    labelString: 'Occurence'
                },
                ticks: {
                  max: 130,
                }
            }],
            xAxes: [{
                scaleLabel: {
                    display: true,
                    labelString: 'UMI Frequency (count)'
                },
                ticks: {
                  fontSize: 12,
                  maxRotation: 90,
                  minRotation: 90,
                  maxTicksLimit: 100,
                }
            }],

        }
    }
});
var slider = document.getElementById("myRange");
var output = document.getElementById("demo");
output.innerHTML = slider.value; // Display the default slider value

// Update the current slider value (each time you drag the slider handle)
slider.oninput = function() {
    output.innerHTML = this.value;
    chart1.options.scales.yAxes[0].ticks.max = parseInt(this.value);
    chart1.update();
}

var ctx2 = document.getElementById('freqChart').getContext('2d');
var chart2 = new Chart(ctx2, {
    // The type of chart we want to create
    type: 'bar',

    // The data for our dataset
    data: {
        labels: mamData_2_umis,
        datasets: [{
            label: "Occurence",
            backgroundColor: 'rgb(99, 99, 255)',
            borderColor: 'rgb(99, 99, 255)',
            data: mamData_2_counts,
        }],

    },

    // Configuration options go here
    options: {
        scales: {
            yAxes: [{
                scaleLabel: {
                    display: true,
                    labelString: 'Count'
                }
            }],
            xAxes: [{
                scaleLabel: {
                    display: true,
                    labelString: 'UMI Sequence'
                },
                ticks: {
                  fontSize: 12,
                  maxRotation: 90,
                  minRotation: 90
                }
            }]
        }
    }
});
var ctx3 = document.getElementById('giniChart').getContext('2d');
var data31 = mamData_3_bins.map((x, i) => {
  return {
    x: x,
    y: mamData_3_result[i]
  };
});
var data32 = mamData_3_bins.map((x, i) => {
  return {
    x: x,
    y: mamData_3_bins[i]
  };
});
var chart3 = new Chart(ctx3, {
    // The type of chart we want to create
    type: 'scatter',

    // The data for our dataset
    data: {
        datasets: [{
            label: "Observed",
            data: data31,
            fill: false,
            borderColor: 'rgb(255, 116, 0)'
          },
          {
            label: "Perfect Equality",
            data: data32,
            fill: false,
            borderColor: 'rgb(28, 85, 255)'
          },
        ],
    },

    // Configuration options go here
    options: {
      animation: false,
        scales: {
            yAxes: [{
                scaleLabel: {
                    display: true,
                    labelString: 'Fraction of Sequencing "Wealth"'
                }
            }],
            xAxes: [{
                scaleLabel: {
                    display: true,
                    labelString: 'Fraction of Gene Population'
                }
            }]
        }
    }
});
</script>

{% endblock %}
